### PR TITLE
Fix opt softmax small nit

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

- Following the PR #18057 we found out that the argument `dtype_attn_weights ` is not defined outside the condition `if attention_mask is not None:` which makes the modeling code prone to potential bugs. Also, the solution provided in #18057 is clearer so this PR addresses the same change to make the implementations consistent across models that suffers from the same Softmax issue. 

cc @ydshieh @sgugger 

Can also confirm slow tests pass!

### CODE REVIEW

**Review by ydshieh — APPROVED** [review]
Thank you @younesbelkada :-)

File: src/transformers/models/opt/modeling_opt.py [hunk 1]
```diff
@@ -218,11 +218,10 @@ def forward(
             attn_weights = attn_weights.view(bsz, self.num_heads, tgt_len, src_len) + attention_mask
             attn_weights = torch.max(attn_weights, torch.tensor(torch.finfo(attn_weights.dtype).min))
             attn_weights = attn_weights.view(bsz * self.num_heads, tgt_len, src_len)
-            dtype_attn_weights = attn_weights.dtype
 
         # upcast to fp32 if the weights are in fp16. Please see https://github.com/huggingface/transformers/pull/17437
-        if dtype_attn_weights == torch.float16:
-            attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(dtype_attn_weights)
+        if attn_weights.dtype == torch.float16:
+            attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(attn_weights.dtype)
```
Comment (younesbelkada) [inline src/transformers/models/opt/modeling_opt.py:224]:
```suggestion
            attn_weights = nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32).to(torch.float16)
```

**Review by sgugger — APPROVED** [review]
Thanks for the fix!
